src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java

```diff
@@ -99,32 +117,68 @@ public PaymentHandleResponse createPayment(@Nonnull String ctpPaymentId) {
+                                            return approvalUrlOpt
+                                                    .map(PaymentHandleResponse::of201CreatedApprovalUrl)
+                                                    .orElseGet(() -> of400BadRequest(
+                                                            format("Payment or cart for ctpPaymentId=[%s] not found", ctpPaymentId)));
+                                        });
                             })
                     // TODO: re-factor compose !!!!
                     .thenCompose(stage -> stage);
-            return executeWithExceptionallyHandling(ctpPaymentId, CTP_PAYMENT_ID, createPaymentCS);
+            return runWithExceptionallyHandling(ctpPaymentId, CTP_PAYMENT_ID, createPaymentCS);
         } catch (Exception e) {
             logger.error("Error while processing payment ID {}", ctpPaymentId, e);
             return of500InternalServerError("See the logs");
         }
     }
 
-    public PaymentHandleResponse patchAddress(@Nonnull Cart cart, @Nonnull String paypalPlusPaymentId) {
+    public PaymentHandleResponse patchAddress(@Nonnull Cart cartWithPaymentsExpansion,
+                                              @Nonnull String paypalPlusPaymentId) {
         try {
-            Payment paypalPlusPayment = new Payment().setId(paypalPlusPaymentId);
-            ShippingAddress shippingAddress = shippingAddressMapper.ctpAddressToPaypalPlusAddress(cart.getShippingAddress());
-            Patch replace = new Patch("add", "/transactions/0/item_list/shipping_address").setValue(shippingAddress);
-            CompletionStage<PaymentHandleResponse> patchCS = paypalPlusFacade.getPaymentService().patch(paypalPlusPayment, replace)
-                    .thenApply(payment -> PaymentHandleResponse.ofHttpStatus(HttpStatus.OK));
-            return executeWithExceptionallyHandling(paypalPlusPaymentId, PAYPAL_PLUS_PAYMENT_ID, patchCS);
+            Optional<String> paymentIdOpt = getCtpPaymentId(cartWithPaymentsExpansion, paypalPlusPaymentId);
+            return paymentIdOpt.map(paymentId -> {
+                Payment paypalPlusPayment = new Payment().setId(paypalPlusPaymentId);
+                ShippingAddress shippingAddress = shippingAddressMapper.ctpAddressToPaypalPlusAddress(cartWithPaymentsExpansion.getShippingAddress());
+                Patch replace = new Patch("add", "/transactions/0/item_list/shipping_address").setValue(shippingAddress);
+                AddInterfaceInteraction addInterfaceInteractionAction = createAddInterfaceInteractionAction(replace, REQUEST);
+                CompletionStage<PaymentHandleResponse> patchCS = ctpFacade.getPaymentService()
+                        // 1. add patch request to the ctp payment interface interaction
+                        .updatePayment(paymentId, Collections.singletonList(addInterfaceInteractionAction))
+                        .thenCompose(payment -> paypalPlusFacade.getPaymentService()
```

Comment:
same as above: too long completion chain with nested stages. Carry out it to some method.

Reply:
https://github.com/commercetools/commercetools-paypal-plus-integration/pull/49/commits/191dad9b40d6ab6ebdd4392d864f93a07e45a8c9